rust/feature-flags/src/flags/flag_matching.rs

```diff
@@ -425,62 +426,192 @@ impl FeatureFlagMatcher {
-            .evaluate_flags_in_level(
-                &feature_flags.flags,
-                &mut evaluated_flags_map,
-                &person_property_overrides,
-                &group_property_overrides,
-                &hash_key_overrides,
+    /// Handles errors during database preparation by creating error responses for affected flags.
+    fn handle_db_preparation_error(
+        &self,
+        flags_requiring_db_preparation: &[&FeatureFlag],
+        error: &FlagError,
+        evaluated_flags_map: &mut HashMap<String, FlagDetails>,
+    ) {
+        let reason = parse_exception_for_prometheus_label(error);
+        evaluated_flags_map.extend(flags_requiring_db_preparation.iter().map(|flag| {
+            (
+                flag.key.clone(),
+                FlagDetails::create_error(flag, reason, None),
             )
-            .await;
-        errors_while_computing_flags |= level_errors;
-        evaluated_flags_map.extend(level_evaluated_flags_map);
+        }));
 
-        FlagsResponse {
-            errors_while_computing_flags,
-            flags: evaluated_flags_map,
-            quota_limited: None,
-            request_id,
-            config: ConfigResponse::default(),
+        error!(
+            "Error preparing flag evaluation state for team {} project {} distinct_id {}: {:?}",
+            self.team_id, self.project_id, self.distinct_id, error
+        );
+
+        inc(
+            FLAG_EVALUATION_ERROR_COUNTER,
+            &[("reason".to_string(), reason.to_string())],
+            1,
+        );
+    }
+
+    /// Evaluates flags using dependency graph to handle flag dependencies correctly.
+    /// Returns true if there were errors during evaluation.
+    async fn evaluate_flags_with_dependency_graph(
+        &mut self,
+        feature_flags: &FeatureFlagList,
+        person_property_overrides: &Option<HashMap<String, Value>>,
+        group_property_overrides: &Option<HashMap<String, HashMap<String, Value>>>,
+        hash_key_overrides: &Option<HashMap<String, String>>,
+        evaluated_flags_map: &mut HashMap<String, FlagDetails>,
+    ) -> bool {
+        let (flag_dependency_graph, errors) =
+            match DependencyGraph::from_nodes(&feature_flags.flags) {
+                Ok((graph, errors)) => (graph, errors),
+                Err(e) => {
+                    self.handle_dependency_graph_error("build feature flag dependency graph", &e);
+                    return true;
+                }
+            };
+
+        let mut errors_while_computing_flags = !errors.is_empty();
+        if errors_while_computing_flags {
+            self.handle_dependency_graph_errors(&errors);
+        }
+
+        let evaluation_stages = match flag_dependency_graph.evaluation_stages() {
+            Ok(stages) => stages,
+            Err(e) => {
+                self.handle_dependency_graph_error("get evaluation stages", &e);
+                debug_assert!(
```

Comment:
Why keep this `debug_assert`?

Reply:
This code path should be unreachable. And my normal preference is to indicate that by calling `unreachable` here. However, should I be wrong about that (for example, maybe someone changes something later that invalidates that contract), I don't want to bring down this service with a panic. I'd rather it fail gracefully.

A `debug_assert` feels like a nice compromise. When developing, we'll get a panic and it will be impossible to ignore. But in production, the service will go on. I'm happy to reconsider. I'll add a comment.